test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java

```diff
@@ -0,0 +1,464 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.action;
+
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import net.openhft.chronicle.core.util.SerializableBiFunction;
+import net.openhft.chronicle.core.util.SerializableConsumer;
+import org.apache.cassandra.db.SystemKeyspace;
+import org.apache.cassandra.dht.IPartitioner;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.distributed.api.ICluster;
+import org.apache.cassandra.distributed.api.IInstance;
+import org.apache.cassandra.distributed.api.IInvokableInstance;
+import org.apache.cassandra.distributed.api.IIsolatedExecutor;
+import org.apache.cassandra.distributed.impl.Instance;
+import org.apache.cassandra.distributed.shared.VersionedApplicationState;
+import org.apache.cassandra.gms.ApplicationState;
+import org.apache.cassandra.gms.EndpointState;
+import org.apache.cassandra.gms.Gossiper;
+import org.apache.cassandra.gms.VersionedValue;
+import org.apache.cassandra.io.util.DataInputBuffer;
+import org.apache.cassandra.io.util.DataOutputBuffer;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.locator.TokenMetadata;
+import org.apache.cassandra.net.MessagingService;
+import org.apache.cassandra.schema.MigrationManager;
+import org.apache.cassandra.service.PendingRangeCalculatorService;
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.utils.FBUtilities;
+
+import static org.apache.cassandra.distributed.impl.DistributedTestSnitch.toCassandraInetAddressAndPort;
+
+
+public class GossipHelper
+{
+
+    public static ClusterAction clusterAction(InstanceAction action)
```

Comment:
ClusterAction seems to be a abstraction over looping, so feel the code is more clear by just looping

examples:

```
clusterAction(statusToBootstrap(newInstance)).apply(cluster);
// vs
cluster.forEach(i -> statusToBootstrap(i, newInstance));
```
 
or use a specific set of nodes

```
clusterAction(disseminateGossipState(newInstance), 1, 2).apply(cluster);
// vs
IntStream.of(1, 2).forEach(i ->  disseminateGossipState(i, newInstance));
```

In these examples, it is more clear to see the loop than the ClusterAction abstraction.

Reply:
Maybe we can do something like this?

```
            cluster.run(statusToBootstrap(newInstance));
            cluster.run(pullSchemaFrom(cluster.get(1)), newInstance.config().num());
            cluster.run(bootstrap(), newInstance.config().num());
```

This could be a good middle ground. You're right that making it all through classes made it look too much like EJB. That said, I still think at least some level of abstraction can be benficial for reuse and making a use of "actions" in Harry: scheduling different clusterwide actions withouot having to write them every time, and being able to reuse them everywhere. WDYT?

Reply:
in your examples there are no real differences between `run` and `forEach`, so I rather have `forEach` only.  An issue we face today is that Instance has a lot of similar, but slightly different methods, which means people mess these up all the time; having two functions that do the same thing worries me about adding extra complexity.

API signatures

```
public abstract class AbstractCluster<I extends IInstance> implements ICluster<I>, AutoCloseable
{
  ...
  public void forEach(Consumer<? super I> consumer)
  ...
  public void run(InstanceAction action)
  ..
  public interface InstanceAction extends Serializable
  {
     void apply(IInvokableInstance target);
  }
```

So the main difference is that we hard code the interface rather than relying on what the cluster provides, and Serializable (but this isn't safe as it works with Instance, which isn't Serializable).  

Also, the current InstancAction has cluster as the first argument, we can always update forEach to offer a BiConsumer

```
public void forEach(BiConsumer<? super ICluster<I>, ? super I> consumer) // note, we can change the signature of AbstractCluster to capture the concrete type, this would make it so you can code against Cluster or UpgradeCluster rather than the interface
```

Reply:
Spoke in slack, I took a stab at removing the interface in favor of functions; branch = https://github.com/dcapwell/cassandra/tree/CASSANDRA-15935-range-movement-improvements, commit = https://github.com/dcapwell/cassandra/commit/ca34aee9ed43a71e26c92f09f6c6e99204816821

Reply:
One thing to notice, `Serializable` goes away in the patch since static methods don't have anything to capture, so the lambda in bootstrap doesn't grab hold of something that isn't `Serializable` so doesn't fail.  It failed with the `BootstrapAction` as the lambda grabbed hold of `this` so required `this` to be `Serializable`; capturing lambdas... joy!

Reply:
> the current InstancAction has cluster as the first argument

`InstanceAction` doesn't have cluster as a first argument, it's just a consumer over invokable. 

>  Serializable goes away in the patch 

I did remove `Serializable` from `InstanceAction`, too. It wasn't strictly required here either, but did made things like capturing multiple variables more convenient. 

> we hard code the interface rather than relying on what the cluster provides

We do not hardcode anything, we're using `run(Consumer<I> action)` same as `foreach` does.

To address a more general point,

> in your examples there are no real differences between run and forEach, so I rather have forEach only.

There are differences:

1. `Action` is an atomic unit of logic, unlike a static method. You can immediately see all things related to a specific action, reuse, and move then at your discression. Using static methods will quickly get out of hand when we have more sophisticated actions.
2. Separation of input arguments (for example "disseminate gossip state of the node X") and `target`, making `target` explicit and common for all cases.
3. Main idea behind the `Action` was to create reusable pieces of logic you can apply to cluster or nodes. For now, logic is simple, like "pull schema fom X, and then bootstrap", but we can reuse similar sequences of steps in: 
   a. Harry, where we can schedule different actions against different sets of nodes, producing reliable results
   b. In upgrade test, where you'll be able to run named actions aginst instances despite the fact they have different versions. 
To do (a) and (b) with static methods, we'll have to _still_ implement some interface. 
4. Finally, this is useful things like chaining actions:
```
            cluster.run(asList(pullSchemaFrom(cluster.get(1)),
                                         bootstrap()),
                        newInstance.config().num());
```

Reply:
Since this conversation is getting lengthier, moving it to jira.